docs/configuration/config.md
# Config Items

> Here we use json with comments for documentation

```json
{
  // ClickHouse config
  "clickhouse": {
    // cluster the ClickHouse node belongs
    "cluster": "test",
    // hosts for connection, it's Array(Array(String))
    // we can put hosts with same shard into the inner array
    // it helps data deduplication for ReplicateMergeTree when driver error occurs
    "hosts": [
      [
        "192.168.101.106",
        "192.168.101.108"
      ],
      [
        "192.168.102.114",
        "192.168.101.110"
      ],
      [
        "192.168.102.115"
      ]
    ],
    "port": 9000,
    "username": "default",
    "password": "",
    // database name
    "db": "default",
    // Whether enable TLS encryption with clickhouse-server
    "secure": false,
    // Whether skip verify clickhouse-server cert if secure=true.
    "insecureSkipVerify": false,
    // retryTimes when error occurs in inserting datas
    "retryTimes": 0,
    // max open connections with each clickhouse node. default to 1.
    "maxOpenConns": 1
  },

  // Kafka config
  "kafka": {
    "brokers": "127.0.0.1:9093",
    
    // jave client style security authentication
    "security":{
        "security.protocol": "SASL_PLAINTEXT",
        "sasl.kerberos.service.name": "kafka",
        "sasl.mechanism":"GSSAPI",
        "sasl.jaas.config":"com.sun.security.auth.module.Krb5LoginModule required useKeyTab=true storeKey=true debug=true keyTab=\"/etc/security/mmmtest.keytab\" principal=\"mmm@ALANWANG.COM\";"
    },

    // SSL
    "tls": {
      "enable": false,
      // Required. It's the CA certificate with which Kafka brokers certs be signed.
      "caCertFiles": "/etc/security/ca-cert",
      // Required if Kafka brokers require client authentication.
      "clientCertFile": "",
      // Required if and only if ClientCertFile is present.
      "clientKeyFile": ""
    },

    // SASL
    "sasl": {
      "enable": false,
      // Mechanism is the name of the enabled SASL mechanism.
      // Possible values: PLAIN, SCRAM-SHA-256, SCRAM-SHA-512, GSSAPI (defaults to PLAIN)
      "mechanism": "PLAIN",
      // Username is the authentication identity (authcid) to present for
      // SASL/PLAIN or SASL/SCRAM authentication
      "username": "",
      // Password for SASL/PLAIN or SASL/SCRAM authentication
      "password": "",
      "gssapi": {
        // authtype - 1. KRB5_USER_AUTH, 2. KRB5_KEYTAB_AUTH
        "authtype": 0,
        "keytabpath": "",
        "kerberosconfigpath": "",
        "servicename": "",
        "username": "",
        "password": "",
        "realm": "",
        "disablepafxfast": false
      }
    },

    // kafka version, if you use sarama, the version must be specified
    "version": "2.5.0"
  },

  "task": {
    "name": "test_dynamic_schema",
    // kafka client, possible values: sarama, kafka-go. (defaults to sarama)
    "kafkaClient": "sarama",
    // kafka topic
    "topic": "topic",
    // kafka consume from earliest or latest
    "earliest": true,
    // kafka consumer group
    "consumerGroup": "group",

    // message parser
    "parser": "json",

    // clickhouse table name
    "tableName": "daily",

    // columns of the table
    "dims": [
      {
        // column name
        "name": "timestamp",
        // column type
        "type": "DateTime"
      },
      {
        "name": "name",
        "type": "String"
      },
      {
        "name": "value",
        "type": "Float32",
        // json field name. This must be specified if it doesn't match with the column name.
        "sourcename": "val"
      }
    ],

    // if it's specified, clickhouse_sinker_nali will detect table schema instead of using the fixed schema given by "dims".
    "autoSchema" : true,
    // these columns will be excluded from the detected table schema. This takes effect only if "autoSchema" is true.
    "excludeColumns": [],

    // (experiment feature) detect new fields and their type, and add columns to the ClickHouse table accordingly. This feature requires parser be "fastjson" or "gjson". New fields' type will be one of: Int64, Float64, String.
    // A column is added for new key K if all following conditions are true:
    // - K isn't in ExcludeColumns
    // - number of existing columns doesn't reach MaxDims-1
    // - WhiteList is empty, or K matchs WhiteList
    // - BlackList is empty, or K doesn't match BlackList
    "dynamicSchema": {
      // whether enable this feature, default to false
      "enable": true,
      // the upper limit of dynamic columns number, <=0 means math.MaxInt16. protecting dirty data attack
      "maxDims": 1024,
      // the regexp of white list. syntax reference: https://github.com/google/re2/wiki/Syntax
      "whiteList": "^[0-9A-Za-z_]+$",
      // the regexp of black list
      "blackList": "@"
    },

    // shardingKey is the column name to which sharding against
    "shardingKey": "",
    // shardingPolicy is `stripe,<interval>`(requires ShardingKey be numerical) or `hash`(requires ShardingKey be string)
    "shardingPolicy": "",

    // interval of flushing the batch. Default to 5, max to 600.
    "flushInterval": 5,
    // batch size to insert into clickhouse. sinker will round upward it to the the nearest 2^n. Default to 262114, max to 1048576.
    "bufferSize": 262114,

    // In the absence of time zone information, interprets the time as in the given location. Default to "Local" (aka /etc/localtime of the machine on which sinker runs)
    "timeZone": "",
    // Time unit when interprete a number as time. Default to 1.0.
    // Java's timestamp is milliseconds since epoch. Change timeUnit to 0.001 at this case.
    "timeUnit": 1.0
  },

  // log level, possible value: "debug", "info", "warn", "error", "dpanic", "panic", "fatal". Default to "info".
  "logLevel": "debug"
}
```


config/config.go
/*Copyright [2019] housepower

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package config

import (
	"encoding/json"
	"io/ioutil"
	"regexp"
	"strings"

	"github.com/forever765/clickhouse_sinker_nali/util"

	"github.com/pkg/errors"
)

// Config struct used for different configurations use
type Config struct {
	Kafka      KafkaConfig
	Clickhouse ClickHouseConfig
	Task       *TaskConfig
	Tasks      []*TaskConfig
	Assignment Assignment
	LogLevel   string
	LogPaths   string
	SinkerListenPort int
	GeoipFilePath	string
}

// KafkaConfig configuration parameters
type KafkaConfig struct {
	Brokers  string
	Version  string
	Security map[string]string
	TLS      struct {
		Enable         bool
		CaCertFiles    string // Required. It's the CA cert.pem with which Kafka brokers certs be signed.
		ClientCertFile string // Required for client authentication. It's client cert.pem.
		ClientKeyFile  string // Required if and only if ClientCertFile is present. It's client key.pem.

		TrustStoreLocation string //JKS format of CA certificate, used to extract CA cert.pem.
		TrustStorePassword string
		KeystoreLocation   string //JKS format of client certificate and key, used to extrace client cert.pem and key.pem.
		KeystorePassword   string
		EndpIdentAlgo      string
	}
	//simplified sarama.Config.Net.SASL to only support SASL/PLAIN and SASL/GSSAPI(Kerberos)
	Sasl struct {
		// Whether or not to use SASL authentication when connecting to the broker
		// (defaults to false).
		Enable bool
		// Mechanism is the name of the enabled SASL mechanism.
		// Possible values: PLAIN, SCRAM-SHA-256, SCRAM-SHA-512, GSSAPI (defaults to PLAIN)
		Mechanism string
		// Username is the authentication identity (authcid) to present for
		// SASL/PLAIN or SASL/SCRAM authentication
		Username string
		// Password for SASL/PLAIN or SASL/SCRAM authentication
		Password string
		GSSAPI   struct {
			AuthType           int //1. KRB5_USER_AUTH, 2. KRB5_KEYTAB_AUTH
			KeyTabPath         string
			KerberosConfigPath string
			ServiceName        string
			Username           string
			Password           string
			Realm              string
			DisablePAFXFAST    bool
		}
	}
}

// ClickHouseConfig configuration parameters
type ClickHouseConfig struct {
	Cluster   string
	DB        string
	Hosts     [][]string
	Port      int
	Username  string
	Password  string
	DsnParams string

	// Whether enable TLS encryption with clickhouse-server
	Secure bool
	// Whether skip verify clickhouse-server cert
	InsecureSkipVerify bool

	RetryTimes   int //<=0 means retry infinitely
	MaxOpenConns int
}

// Task configuration parameters
type TaskConfig struct {
	Name string

	KafkaClient   string
	Topic         string
	ConsumerGroup string

	// Earliest set to true to consume the message from oldest position
	Earliest bool
	Parser   string
	// the csv cloum title if Parser is csv
	CsvFormat []string
	Delimiter string

	TableName string

	// AutoSchema will auto fetch the schema from clickhouse
	AutoSchema     bool
	ExcludeColumns []string
	Dims           []struct {
		Name       string
		Type       string
		SourceName string
	} `json:"dims"`
	// DynamicSchema will add columns present in message to clickhouse. Requires AutoSchema be true.
	DynamicSchema struct {
		Enable  bool
		MaxDims int // the upper limit of dynamic columns number, <=0 means math.MaxInt16. protecting dirty data attack
		// A column is added for new key K if all following conditions are true:
		// - K isn't in ExcludeColumns
		// - number of existing columns doesn't reach MaxDims-1
		// - WhiteList is empty, or K matchs WhiteList
		// - BlackList is empty, or K doesn't match BlackList
		WhiteList string // the regexp of white list
		BlackList string // the regexp of black list
	}
	// PrometheusSchema expects each message is a Prometheus metric(timestamp, value, metric name and a list of labels).
	PrometheusSchema bool

	// ShardingKey is the column name to which sharding against
	ShardingKey string `json:"shardingKey,omitempty"`
	// ShardingPolicy is `stripe,<interval>`(requires ShardingKey be numerical) or `hash`(requires ShardingKey be string)
	ShardingPolicy string `json:"shardingPolicy,omitempty"`

	FlushInterval int     `json:"flushInterval,omitempty"`
	BufferSize    int     `json:"bufferSize,omitempty"`
	TimeZone      string  `json:"timeZone"`
	TimeUnit      float64 `json:"timeUnit"`
	GeoipHandle	bool
	AutoUpdateGeoIPDB	string
}

type Assignment struct {
	Version   int
	UpdatedAt int64               // timestamp when created
	UpdatedBy string              // leader instance
	Map       map[string][]string // map instance to a list of task_name
}

const (
	MaxBufferSize             = 1 << 20 //1048576
	defaultBufferSize         = 1 << 18 //262144
	maxFlushInterval          = 600
	defaultFlushInterval      = 5
	defaultGeoipHandle        = false
	defaultTimeZone           = "Local"
	defaultLogLevel           = "info"
	defaultKerberosConfigPath = "/etc/krb5.conf"
	defaultMaxOpenConns       = 1
)

func ParseLocalCfgFile(cfgPath string) (cfg *Config, err error) {
	cfg = &Config{}
	var b []byte
	b, err = ioutil.ReadFile(cfgPath)
	if err != nil {
		err = errors.Wrapf(err, "")
		return
	}
	if err = json.Unmarshal(b, cfg); err != nil {
		err = errors.Wrapf(err, "")
		return
	}
	return
}

// normallize and validate configuration
func (cfg *Config) Normallize() (err error) {
	if len(cfg.Clickhouse.Hosts) == 0 || cfg.Kafka.Brokers == "" {
		err = errors.Errorf("invalid configuration")
		return
	}
	if cfg.Kafka.Version == "" {
		cfg.Kafka.Version = "2.5.0"
	}

	cfg.convertKfkSecurity()
	if cfg.Kafka.Sasl.Enable {
		cfg.Kafka.Sasl.Mechanism = strings.ToUpper(cfg.Kafka.Sasl.Mechanism)
		switch cfg.Kafka.Sasl.Mechanism {
		case "PLAIN", "SCRAM-SHA-256", "SCRAM-SHA-512", "GSSAPI":
		default:
			err = errors.Errorf("kafka SASL mechanism %s is unsupported", cfg.Kafka.Sasl.Mechanism)
			return
		}
	}
	if cfg.Clickhouse.RetryTimes < 0 {
		cfg.Clickhouse.RetryTimes = 0
	}
	if cfg.Clickhouse.MaxOpenConns <= 0 {
		cfg.Clickhouse.MaxOpenConns = defaultMaxOpenConns
	}

	if cfg.Task != nil {
		cfg.Tasks = append(cfg.Tasks, cfg.Task)
		cfg.Task = nil
	}
	for _, taskCfg := range cfg.Tasks {
		if err = cfg.normallizeTask(taskCfg); err != nil {
			return
		}
	}
	switch strings.ToLower(cfg.LogLevel) {
	case "debug", "info", "warn", "error", "dpanic", "panic", "fatal":
	default:
		cfg.LogLevel = defaultLogLevel
	}
	return
}

func (cfg *Config) normallizeTask(taskCfg *TaskConfig) (err error) {
	if taskCfg.KafkaClient == "" || (cfg.Kafka.Sasl.Enable && cfg.Kafka.Sasl.Username == "") {
		// known limitations of kafka-go:
		// - The Reader API is too high-level. There's no generation cleanup callback which sarama provides.
		// - Doesn't support SASL/GSSAPI(Kerberos). https://github.com/segmentio/kafka-go/issues/539
		taskCfg.KafkaClient = "sarama"
	}
	if taskCfg.Parser == "" || taskCfg.Parser == "json" {
		taskCfg.Parser = "fastjson"
	}

	for i := range taskCfg.Dims {
		if taskCfg.Dims[i].SourceName == "" {
			taskCfg.Dims[i].SourceName = util.GetSourceName(taskCfg.Dims[i].Name)
		}
	}

	if taskCfg.FlushInterval <= 0 {
		taskCfg.FlushInterval = defaultFlushInterval
	} else if taskCfg.FlushInterval > maxFlushInterval {
		taskCfg.FlushInterval = maxFlushInterval
	}
	if taskCfg.BufferSize <= 0 {
		taskCfg.BufferSize = defaultBufferSize
	} else if taskCfg.BufferSize > MaxBufferSize {
		taskCfg.BufferSize = MaxBufferSize
	} else {
		taskCfg.BufferSize = 1 << util.GetShift(taskCfg.BufferSize)
	}
	if taskCfg.TimeZone == "" {
		taskCfg.TimeZone = defaultTimeZone
	}
	if taskCfg.TimeUnit == 0.0 {
		taskCfg.TimeUnit = float64(1.0)
	}
	if taskCfg.PrometheusSchema {
		taskCfg.DynamicSchema.Enable = true
		taskCfg.AutoSchema = true
	}
	if taskCfg.DynamicSchema.Enable {
		if taskCfg.Parser != "fastjson" && taskCfg.Parser != "gjson" {
			err = errors.Errorf("Parser %s doesn't support DynamicSchema", taskCfg.Parser)
			return
		}
	}
	if taskCfg.DynamicSchema.WhiteList != "" {
		if _, err = regexp.Compile(taskCfg.DynamicSchema.WhiteList); err != nil {
			err = errors.Wrapf(err, "WhiteList %s is invalid regexp", taskCfg.DynamicSchema.WhiteList)
			return
		}
	}
	if taskCfg.DynamicSchema.BlackList != "" {
		if _, err = regexp.Compile(taskCfg.DynamicSchema.BlackList); err != nil {
			err = errors.Wrapf(err, "BlackList %s is invalid regexp", taskCfg.DynamicSchema.BlackList)
			return
		}
	}
	// if GeoipHandle not set, don't open it
	if !taskCfg.GeoipHandle {
		taskCfg.GeoipHandle = defaultGeoipHandle
	}
	// Auto update geoip db file cron job
	if taskCfg.AutoUpdateGeoIPDB != "" {
		util.AddUpdateCronTask(taskCfg.AutoUpdateGeoIPDB)
	}
	return
}

//convert java client style configuration into sinker
func (cfg *Config) convertKfkSecurity() {
	if protocol, ok := cfg.Kafka.Security["security.protocol"]; ok {
		if strings.Contains(protocol, "SASL") {
			cfg.Kafka.Sasl.Enable = true
		}
		if strings.Contains(protocol, "SSL") {
			cfg.Kafka.TLS.Enable = true
		}
	}

	if cfg.Kafka.TLS.Enable {
		if endpIdentAlgo, ok := cfg.Kafka.Security["ssl.endpoint.identification.algorithm"]; ok {
			cfg.Kafka.TLS.EndpIdentAlgo = endpIdentAlgo
		}
		if trustStoreLocation, ok := cfg.Kafka.Security["ssl.truststore.location"]; ok {
			cfg.Kafka.TLS.TrustStoreLocation = trustStoreLocation
		}
		if trustStorePassword, ok := cfg.Kafka.Security["ssl.truststore.password"]; ok {
			cfg.Kafka.TLS.TrustStorePassword = trustStorePassword
		}
		if keyStoreLocation, ok := cfg.Kafka.Security["ssl.keystore.location"]; ok {
			cfg.Kafka.TLS.KeystoreLocation = keyStoreLocation
		}
		if keyStorePassword, ok := cfg.Kafka.Security["ssl.keystore.password"]; ok {
			cfg.Kafka.TLS.KeystorePassword = keyStorePassword
		}
	}
	if cfg.Kafka.Sasl.Enable {
		if mechanism, ok := cfg.Kafka.Security["sasl.mechanism"]; ok {
			cfg.Kafka.Sasl.Mechanism = mechanism
		}
		if config, ok := cfg.Kafka.Security["sasl.jaas.config"]; ok {
			configMap := readConfig(config)
			if strings.Contains(cfg.Kafka.Sasl.Mechanism, "SCRAM") {
				// SCRAM-SHA-256 or SCRAM-SHA-512
				if username, ok := configMap["username"]; ok {
					cfg.Kafka.Sasl.Username = username
				}
				if password, ok := configMap["password"]; ok {
					cfg.Kafka.Sasl.Password = password
				}
			}
			if strings.Contains(cfg.Kafka.Sasl.Mechanism, "GSSAPI") {
				// GSSAPI
				if useKeyTab, ok := configMap["useKeyTab"]; ok {
					if useKeyTab == "true" {
						cfg.Kafka.Sasl.GSSAPI.AuthType = 2
					} else {
						cfg.Kafka.Sasl.GSSAPI.AuthType = 1
					}
				}
				if cfg.Kafka.Sasl.GSSAPI.AuthType == 1 {
					//Username and password
					if username, ok := configMap["username"]; ok {
						cfg.Kafka.Sasl.GSSAPI.Username = username
					}
					if password, ok := configMap["password"]; ok {
						cfg.Kafka.Sasl.GSSAPI.Password = password
					}
				} else {
					//Keytab
					if keyTab, ok := configMap["keyTab"]; ok {
						cfg.Kafka.Sasl.GSSAPI.KeyTabPath = keyTab
					}
					if principal, ok := configMap["principal"]; ok {
						username := strings.Split(principal, "@")[0]
						realm := strings.Split(principal, "@")[1]
						cfg.Kafka.Sasl.GSSAPI.Username = username
						cfg.Kafka.Sasl.GSSAPI.Realm = realm
					}
					if servicename, ok := cfg.Kafka.Security["sasl.kerberos.service.name"]; ok {
						cfg.Kafka.Sasl.GSSAPI.ServiceName = servicename
					}
					if cfg.Kafka.Sasl.GSSAPI.KerberosConfigPath == "" {
						cfg.Kafka.Sasl.GSSAPI.KerberosConfigPath = defaultKerberosConfigPath
					}
				}
			}
		}
	}
}

func (cfg *Config) IsAssigned(instance, task string) (assigned bool) {
	if taskNames, ok := cfg.Assignment.Map[instance]; ok {
		for _, taskName := range taskNames {
			if taskName == task {
				assigned = true
				return
			}
		}
	}
	return
}

func readConfig(config string) map[string]string {
	configMap := make(map[string]string)
	config = strings.TrimSuffix(config, ";")
	fields := strings.Split(config, " ")
	for _, field := range fields {
		if strings.Contains(field, "=") {
			key := strings.Split(field, "=")[0]
			value := strings.Split(field, "=")[1]
			value = strings.Trim(value, "\"")
			configMap[key] = value
		}
	}
	return configMap
}
